### DISCUSSION TITLE
Plot mulitple grains datasets on a single histogramn

### DISCUSSION THREAD

**Discussion by nicholasbyres** [mtex-toolbox/mtex#710] [Ask Anything]
Hi MTEXers,

I am trying to produce a figure that compares the grainsize distribution of several grain2d datasets on one histogram. If I use the simple method of
```matlab
hist(grains_1)
hold on
hist(grains_2)
```
etc. The bars plot overlap one another. However, in some older dodumentation that I came across, for the command plotAngleDistribution (that I can't seem to find now), the bars of the multiple histogram datasets are plotted next to one another that allows for nice comparisson. Is there a simple way to do this for the hist function?

Thanks again and I hope you're all well!

Best,
Nick

**Comment by nyyssont** [comment]
Perhaps subplot is what you are after? Set equal limits for both y-axes and comparison should be easy.
https://se.mathworks.com/help/matlab/ref/subplot.html

Another option would be to extract the bin centers and bin heights from histcounts and plot them as bar plots:
https://se.mathworks.com/help/matlab/ref/histcounts.html
https://se.mathworks.com/help/matlab/ref/bar.html

A third option would be to use "histogram" instead of hist. The plots still overlap, but they are also transparent so can be compared relatively easily:
https://se.mathworks.com/help/matlab/ref/matlab.graphics.chart.primitive.histogram.html

**Reply by nicholasbyres** [reply]
Hi Tuomo,

Subplots wouldn't work as I have too many variables I'm trying to compare, same with changing the FaceAlpha with the histogram command. Extracting the bin centres and heights seems promising though so I'll give that a go.

Many thanks!
Nick

**Reply by kilir** [reply]
Hi Nick,
you can use Matlabs `histcounts` together with the `bar` plot. The latter will take care of the arrangement of the bars in each class if there are multiples. 
May I ask why you'd opt for the number proportion (frequency) of some measure of grain size (d_eq?, or area...) and not an area weighted distribution?

Also there's `grain2d/histogram` which will give you by default the area weighted histogram as well as a handle for the bar plot.
 
Cheers,
Rüdiger

**Reply by nicholasbyres** [reply]
Hi Rüdiger,
thanks for the reply. Yes I am actually interested in the weighted distribution. I realised this and thought I'd have to work around it using ```histcounts``` and ```bar``` plot with the code from the ```grain2d/histogram``` command. This is indeed challenging my coding expertise (which is limited). 

I believe what I want is the  ```grain2d/histogram``` to be able to plot multiple datasets like in ```bar``` plot. Does this make sense?
Cheers,
Nick

**Reply by kilir** [reply]
Yes, totally.
Here's the easier way:
a) make a histogram of frequencies of e.g. equivalent diameter with specified bins using `histcounts` in e.g. fractions
b) multiply each bin population by (d/2)^2 for area weighting or (d/2)^3 for volume weighting
c) normalize again each weighted bin by the sum of the weighted histogram
Repeat with all your samples and plot with `bar`.
This is a reasonably good approximation for most well behaved distributions.

Cheers,
Rüdiger

**Reply by nicholasbyres** [reply]
Hi Rüdiger,
Thanks for your help thus far. The following code was taken from the hist.m file and adapted for use on a single phase.

```
cumArea_TAFR = zeros(numel(binsC)-1,1);
[~,~,binIdC] = histcounts(equivalentRadius(grains_TAFR),binsC);
  % compute the sum of areas belonging to the same bin
  for i = 1:numel(binsC)-1
      cumArea_TAFR(i) = sum(equivalentRadius(grains_TAFR(binIdC == i))) ./ sum(equivalentRadius(grains_RT));
  end
```

The issue here I think is the normalisation as the sum of the bins doesn't total one. I think the issue is dividing by the sum of the equivalentRadius. How can I get around this issue?

Thanks again!
Nick

**Comment by bjoerneske** [comment]
Dear all

With the `parent',axpos{i}`   command you can do a lot.  I do something
like this to plot multiple axes in on figure. This depends in my case of a
cell structure with the grains from different regions:
```matlab
%% Calcite grain sizes in the regions
%extract grains from from the 5 first regions different regions
n =length(mergedGrains_regions)
EQR_cc=cell(n,1);
grainsArea_cc=cell(n,1);
axpos=cell(n,1);
for i =1:n
grains_Area=mergedGrains_regions{i};
grainsArea_cc{i}=grains_Area('calcite');
EQR_cc{i}=grainsArea_cc{i}.equivalentRadius;
end
max_EQR_cc=max(cell2mat(EQR_cc));
% this is the figure
figure
nbins=50;
% you need to define number of columns and rows to make things look nice
mtexFig=newMtexFigure('nrows',n,'ncols',1,'figSize','large');mtexFig.layoutMode='user'
for i =1:n
axpos{i} =subplot(n,1,i);
nbins1=nbins*max(EQR_cc{i})/max_EQR_cc
hist(grainsArea_cc{i},EQR_cc{i},nbins1,'parent',axpos{i})
ax =gca;
ax.Legend.String=['Calcite region' num2str(i) ', n = '
num2str(length(grainsArea_cc{i}))]
end
% just to remove labels and have same max on all axes
maxsize =max(grains('calcite').equivalentRadius);
for i=1:n
mtexFig.children(i).XLim=[0 maxsize];
mtexFig.children(i).FontSize=8;
if i<n
mtexFig.children(i).XTickLabel=[];
end
end
```
man. 25. jan. 2021, 13:09 skrev nicholasbyres <notifications@github.com>:
